### PR TITLE
Autocomplete

### PR DESCRIPTION
**Pull request by Bernardstanislas**

### CODE REVIEW

File: common/autocomplete/awesomplete.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+ * @type {Object}
+ */
+let Autocomplete = {
+    /**
+     * Component will mount.
+     * Check if the Awesomplete library is in the Window object.
+     */
+    componentWillMount() {
+        // Check if Awesomplete is set in Window
+        if (!window.Awesomplete) {
```
Comment (pierr) [inline common/autocomplete/awesomplete.js]: Maybe add the awesomeplete as optional dependencies https://docs.npmjs.com/files/package.json#optionaldependencies

File: common/autocomplete/awesomplete.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            pickList: [],
+            validate: true
+        };
+    },
+    /**
+     * Prop validation
+     * @type {Object}
+     */
+    propTypes: {
+        code: types('string'),
```
Comment (pierr) [inline common/autocomplete/awesomplete.js]: Add documentation on the props here

File: common/autocomplete/awesomplete.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        selectionHandler: types('function'),
+        validate: types('bool')
+    },
+    /**
+     * Initial state.
+     * Retrieve the value from the provided code and pick list.
+     * @return {Object} initial state
+     */
+    getInitialState() {
+        const {code, pickList} = this.props;
```
Comment (pierr) [inline common/autocomplete/awesomplete.js]: :+1:

File: common/autocomplete/awesomplete.js [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+     */
+    _extractListFromData(data) {
+        return data.map(datum => datum.value);
+    },
+    /**
+     * Get code from value in the pick list
+     * @param  {String} value the value
+     * @return {String} the code
+     */
+    _getCodeFromValue(value) {
```
Comment (pierr) [inline common/autocomplete/awesomplete.js]: it seems value is always read from the state

File: common/autocomplete/awesomplete.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        this.setState({value});
+        if (this._changeTimeout) {
+            clearTimeout(this._changeTimeout);
+        }
+        this._changeTimeout = setTimeout(() => {
+            let {inputChangeHandler} = this.props;
+            if (inputChangeHandler) {
+                inputChangeHandler(value);
+            }
+        }, 200);
```
Comment (pierr) [inline common/autocomplete/awesomplete.js]: add a props

File: common/autocomplete/awesomplete.js [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        let {_onInputBlur, _onInputChange} = this;
+        return (
+            <input onBlur={_onInputBlur} onChange={_onInputChange} ref='input' value={value}/>
+        );
+    },
+    /**
+     * Render the consult mode
+     * @return {HTML} rendered element
+     */
+    _renderConsult() {
```
Comment (pierr) [inline common/autocomplete/awesomplete.js]: See if there is two separate components, for now it seems right to let the component render the consult itself.

Comment (pierr) [inline common/autocomplete/awesomplete.js]: Maybe it should be the role of the autoCompleteFor

File: common/autocomplete/example/lib/awesomplete.css [hunk 7]
```diff
@@ -0,0 +1,97 @@
+[hidden] { display: none; }
```
Comment (pierr) [inline common/autocomplete/example/lib/awesomplete.css]: is there a CDN ? It seems there is none.

File: common/autocomplete/example/lib/awesomplete.js [hunk 8]
```diff
@@ -0,0 +1,391 @@
+/**
```
Comment (pierr) [inline common/autocomplete/example/lib/awesomplete.js]: CDN

File: common/autocomplete/example/lib/awesomplete.min.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+"// Awesomplete - Lea Verou - MIT license"
```
Comment (pierr) [inline common/autocomplete/example/lib/awesomplete.min.js]: lopez

File: common/autocomplete/field.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    },
+    /**
+     * Render the component
+     * @return {HTML} the rendered component
+     */
+    render() {
+        let {AutocompleteComponent, isEdit, value} = this.props;
+        let {pickList} = this.state;
+        return (
+            <AutocompleteComponent
```
Comment (pierr) [inline common/autocomplete/field.js]: Maybe the consult / edit could be handled here...

File: common/autocomplete/awesomplete.js (hunk 1 above)
Comment (Bernardstanislas) [inline common/autocomplete/awesomplete.js]: Not on NPM, the only package is not really usable

File: common/autocomplete/awesomplete.js (hunk 4 above)
Comment (Bernardstanislas) [inline common/autocomplete/awesomplete.js]: At least, it is closer to a pure function, with less external calls

File: common/autocomplete/example/lib/awesomplete.css (hunk 7 above)
Comment (Bernardstanislas) [inline common/autocomplete/example/lib/awesomplete.css]: negative..

**Pull request merged**